### PR TITLE
tree: Move logic from proxy binding to node kernel

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

A lot of the proxyBinding logic makes more sense to be part of node kernel, and just use properties on there instead of maps.

There is still some more cleanup to do to finish this migration, this this first step moves most of the logic.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/shared-tree/treeApi.ts [hunk 1]
```diff
@@ -436,6 +436,7 @@ export function runTransaction<
 		const node = treeOrNode as TNode;
 		const t = transaction as (node: TNode) => TResult | typeof rollback;
 		const context = getOrCreateInnerNode(node).context;
+		// TODO: this can fail for unhydrated nodes
```
Comment (taylorsw04) [inline packages/dds/tree/src/shared-tree/treeApi.ts:439]: file an item plz

File: packages/dds/tree/src/simple-tree/proxyBinding.ts [hunk 2]
```diff
@@ -3,97 +3,17 @@
...
-} from "../feature-libraries/index.js";
-import { fail } from "../util/index.js";
-import type { TreeNode } from "./core/index.js";
-// TODO: decide how to deal with dependencies on flex-tree implementation.
-// eslint-disable-next-line import/no-internal-modules
-import { makeTree } from "../feature-libraries/flex-tree/lazyNode.js";
+import type { InnerNode, TreeNode } from "./core/index.js";
 import { getKernel } from "./core/index.js";
 
 // This file contains various maps and helpers for supporting associating simple TreeNodes with their InnerNodes, and swapping those InnerNodes as part of hydration.
```
Comment (taylorsw04) [inline packages/dds/tree/src/simple-tree/proxyBinding.ts:9]: This comment should be updated. However, perhaps this file should also be deleted? Is this helper really useful?

File: packages/dds/tree/src/simple-tree/treeNodeValid.ts [hunk 3]
```diff
@@ -157,8 +157,8 @@ export abstract class TreeNodeValid<TInput> extends TreeNode {
 		);
 
 		const result = schema.prepareInstance(this, node);
-		new TreeNodeKernel(result, schema);
-		setInnerNode(result, node);
+		new TreeNodeKernel(result, schema, node);
```
Comment (taylorsw04) [inline packages/dds/tree/src/simple-tree/treeNodeValid.ts:160]: I know you didn't introduce this, but a floating constructor call makes me want to ask if this is code smell.

File: packages/dds/tree/src/simple-tree/proxyBinding.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxyBinding.ts:9]: There are 39 references to getOrCreateInnerNode, so I punted on dealing with it as part of this PR.

File: packages/dds/tree/src/simple-tree/treeNodeValid.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/treeNodeValid.ts:160]: I don't really like it, but there isn't a clear more robust API I could think of (I considered several options when I originally did this), and this is literally the only place that class will ever be instanced, so it seems fine though odd.

Given pretty much all the other code in here is deceptively fragile and kinda horrifying (ex: preceding line does static method virtual dispatch on a base class blind casted from the this object's constructor parameter to produce whats usally a proxy which is returned from the constructor to impersonate one of the base classes.....) having one scary looking thing thats actually totally safe and robust next to all the actually scary things might not be too bad.

Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/treeNodeValid.ts:163]: I have add a comment to indicate this was intended.

File: packages/dds/tree/src/simple-tree/proxyBinding.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxyBinding.ts]: Old file comment removed.

File: packages/dds/tree/src/shared-tree/treeApi.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/shared-tree/treeApi.ts]: done.

**Review by taylorsw04 — APPROVED** [review]

**Pull request merged**